### PR TITLE
Add the layers drop_path.py, layer_scale.py, mlp.py, swiglu_ffn.py of DINOV2

### PR DESCRIPTION
**Pull request by mohamedamrfaroukebrahim**

### CODE REVIEW

File: paz/models/foundation/dinov2/drop_path.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import keras
+from keras import ops
+from typing import Optional
+
+
+def drop_path_keras_functional(
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/drop_path.py]: We don't need to write "keras"

File: paz/models/foundation/dinov2/drop_path.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    x: keras.KerasTensor, drop_prob: float = 0.0, training: bool = False
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/drop_path.py]: We are not using type annotations.

Comment (oarriaga) [inline paz/models/foundation/dinov2/drop_path.py]: dropout_rate

File: paz/models/foundation/dinov2/drop_path.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    keep_prob = 1.0 - drop_prob
+
+    shape_x = ops.shape(x)
+    shape_for_random = [shape_x[0]] + [1] * (len(shape_x) - 1)
+
+    # random_tensor = ops.random.bernoulli(shape_for_random, p=keep_prob, dtype=x.dtype)
+    random_tensor = keras.random.binomial(
+        shape=shape_for_random, counts=1.0, probabilities=keep_prob, dtype=x.dtype
+    )
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/drop_path.py]: Wrap this as a function that just does bernoulli.

File: paz/models/foundation/dinov2/mlp.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import keras
+from typing import Optional, Type
+
+
+class Mlp(keras.Model):
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/mlp.py]: Acronyms should always be capitalized.

**Review by oarriaga — CHANGES_REQUESTED** [review]

File: paz/models/foundation/dinov2/layers/attention.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+from keras.layers import Dense, Dropout, Layer
+
+
+class Attention(Layer):
+    def __init__(
+        self,
+        dimension,
+        num_heads=8,
+        qkv_bias=False,
+        projection_bias=True,
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: Let's indicate that they are boolean.

File: paz/models/foundation/dinov2/layers/attention.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.dimension = dimension
+        self.num_heads = num_heads
+        self.qkv_bias = qkv_bias
+        self.projection_bias = projection_bias
+        self.attention_drop_rate_val = attention_drop_rate
+        self.projection_drop_rate_val = projection_drop_rate
+        self.head_dim = self.dimension // self.num_heads
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: Let's always use full variable names "dim" for "dimension"

File: paz/models/foundation/dinov2/layers/attention.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        projection_drop_rate=0.0,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.dimension = dimension
+        self.num_heads = num_heads
+        self.qkv_bias = qkv_bias
+        self.projection_bias = projection_bias
+        self.attention_drop_rate_val = attention_drop_rate
+        self.projection_drop_rate_val = projection_drop_rate
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: Let's keep the same variable name and not change it internally in the class.

File: paz/models/foundation/dinov2/layers/attention.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        self.head_dim = self.dimension // self.num_heads
+        self.scale = self.head_dim**-0.5
+
+        initializer = keras.initializers.TruncatedNormal(stddev=0.02)
+        self.qkv = Dense(
+            self.dimension * 3, use_bias=self.qkv_bias, kernel_initializer=initializer, name="qkv"
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: Maybe to make it fit in one line, we could do kwargs = {"kernel_initializer": initializer} and pass the kwargs to both dense functions.

File: paz/models/foundation/dinov2/layers/attention.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        )
+        self.proj = Dense(
+            self.dimension, use_bias=self.projection_bias, kernel_initializer=initializer, name="proj"
+        )
+        self.attention_drop = Dropout(self.attention_drop_rate_val)
+        self.projection_drop = Dropout(self.projection_drop_rate_val)
+
+    def build(self, input_shape):
+        super().build(input_shape)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: Let's try to remove it, since we are not doing anything with it.

File: paz/models/foundation/dinov2/layers/attention.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+    def call(self, x, attn_bias=None, training=None):
+        B, N, C = ops.shape(x)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: Let's rename this to full variable names. batch_size or batch, ...

File: paz/models/foundation/dinov2/layers/attention.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        qkv = self.qkv(x)
+        qkv = ops.reshape(qkv, (B, N, 3, self.num_heads, self.head_dim))
+        qkv = ops.transpose(qkv, axes=(2, 0, 3, 1, 4))
+        q, k, v = qkv[0], qkv[1], qkv[2]
+
+        attn = ops.matmul(q, ops.transpose(k, axes=[0, 1, 3, 2])) * self.scale
+
+        if attn_bias is not None:
+            attn += attn_bias
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: Let's be more explicit: attention = attention + attention_bias

File: paz/models/foundation/dinov2/layers/attention.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        )
+        self.attention_drop = Dropout(self.attention_drop_rate_val)
+        self.projection_drop = Dropout(self.projection_drop_rate_val)
+
+    def build(self, input_shape):
+        super().build(input_shape)
+
+    def call(self, x, attn_bias=None, training=None):
+        B, N, C = ops.shape(x)
+        qkv = self.qkv(x)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: Maybe we should be able to read this as "self.qkv" -> "self.predict_query_key_value"

File: paz/models/foundation/dinov2/layers/attention.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        x = ops.reshape(x, (B, N, C))
+
+        x = self.proj(x)
+        x = self.projection_drop(x, training=training)
+        return x
+
+
+class MemEffAttention(Attention):
+    def call(self, x, attn_bias=None, training=None):
+        return super().call(x, attn_bias=attn_bias, training=training)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/attention.py]: We should try to remove this.

File: paz/models/foundation/dinov2/layers/__init__.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+from .dino_head import DINOHead
+from .layer_scale import LayerScale
+from .mlp import MLP
+from .patch_embed import PatchEmbed
+from .swiglu_ffn import SwiGLUFFN, SwiGLUFFNFused, SwiGLUFFNAligned
+from .block import NestedTensorBlock, CausalAttentionBlock
+from .attention import Attention, MemEffAttention
+from .drop_path import DropPath
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/__init__.py]: Let's do the imports always from root:
"from paz.models.foundation.layers import DINOHead"

File: paz/models/foundation/dinov2/layers/block.py [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+    residual_func,
+    sample_drop_ratio=0.0,
+):
+    b = ops.shape(x)[0]
+
+    subset_float = ops.multiply(ops.cast(b, "float32"), (1.0 - sample_drop_ratio))
+    sample_subset_size = ops.maximum(ops.cast(subset_float, "int32"), 1)
+
+    all_indices = ops.arange(b, dtype="int32")
+    brange = keras.random.shuffle(all_indices)[:sample_subset_size]
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: Rename to something that conveys more meaning to the reader.

File: paz/models/foundation/dinov2/layers/block.py [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+
+
+def drop_add_residual_stochastic_depth(
+    x,
+    residual_func,
+    sample_drop_ratio=0.0,
+):
+    b = ops.shape(x)[0]
+
+    subset_float = ops.multiply(ops.cast(b, "float32"), (1.0 - sample_drop_ratio))
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: why is this a subset

File: paz/models/foundation/dinov2/layers/block.py [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+    sample_subset_size = ops.maximum(ops.cast(subset_float, "int32"), 1)
+
+    all_indices = ops.arange(b, dtype="int32")
+    brange = keras.random.shuffle(all_indices)[:sample_subset_size]
+
+    x_subset = ops.take(x, brange, axis=0)
+
+    residual = residual_func(x_subset)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: apply_residual

File: paz/models/foundation/dinov2/layers/block.py [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+        residual_flat = ops.reshape(residual, (ops.shape(residual)[0], -1))
+
+        scaled_residual = ops.multiply(residual_flat, residual_scale_factor)
+
+        existing_values = ops.take(x_flat, brange, axis=0)
+        updated_values = ops.add(existing_values, scaled_residual)
+        scatter_indices = ops.expand_dims(brange, axis=1)
+        x_plus_residual_flat = ops.scatter_update(x_flat, scatter_indices, updated_values)
+
+        return ops.reshape(x_plus_residual_flat, original_shape)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: let's just have a single return

File: paz/models/foundation/dinov2/layers/block.py [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+        for j in range(num_blocks):
+            rows = seqlens[i]
+            cols = seqlens[j]
+            if i == j:
+                block = ops.ones((rows, cols), dtype=dtype)
+            else:
+                block = ops.zeros((rows, cols), dtype=dtype)
+            block_row.append(block)
+        rows_of_blocks.append(ops.concatenate(block_row, axis=1))
+    return ops.concatenate(rows_of_blocks, axis=0)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: Let's try to understand what this function is doing and change the name to show me that.

File: paz/models/foundation/dinov2/layers/block.py [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+
+
+def get_attn_bias_and_cat(x_list, branges=None):
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: same as before let's try to understand this.

File: paz/models/foundation/dinov2/layers/block.py [hunk 21]
```diff
@@ -0,0 +1,379 @@
...
+        self.ffn_norm.gamma.assign(ops.ones(self.ffn_norm.gamma.shape))
+        self.ffn_norm.beta.assign(ops.zeros(self.ffn_norm.beta.shape))
+
+        self.feed_forward.fc1.kernel.assign(
+            keras.random.normal(self.feed_forward.fc1.kernel.shape, stddev=init_fc_std)
+        )
+        self.feed_forward.fc2.kernel.assign(
+            keras.random.normal(self.feed_forward.fc2.kernel.shape, stddev=init_proj_std)
+        )
+
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: Let's try to encapsulate this into a function that tells me what is happening. We can then try to use other functions such as initializers?

File: paz/models/foundation/dinov2/layers/block.py [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+                sample_drop_ratio=self.sample_drop_ratio,
+                scaling_vector=self.ls1.gamma if isinstance(self.ls1, LayerScale) else None,
+            )
+            x_list = drop_add_residual_stochastic_depth_list(
+                x_list,
+                residual_func=ffn_residual_func,
+                sample_drop_ratio=self.sample_drop_ratio,
+                scaling_vector=self.ls2.gamma if isinstance(self.ls2, LayerScale) else None,
+            )
+            return x_list
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: one return

File: paz/models/foundation/dinov2/layers/block.py [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+
+        else:
+
+            def attn_residual_func(x, attn_bias=None):
+                return self.ls1(self.attn(self.norm1(x), attn_bias=attn_bias))
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: I don't know what ls1 is. Let's use full variable names of attention, normalization, etc.

File: paz/models/foundation/dinov2/layers/block.py [hunk 24]
```diff
@@ -0,0 +1,379 @@
...
+
+            def ffn_residual_func(x, attn_bias=None):
+                return self.ls2(self.mlp(self.norm2(x)))
+
+            attn_mask, x_cat = get_attn_bias_and_cat(x_list)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: let's use full variable names. Let's not use cat.

File: paz/models/foundation/dinov2/layers/block.py [hunk 25]
```diff
@@ -0,0 +1,379 @@
...
+
+            return output_list
+
+    def call(self, x_or_x_list, training=None):
+        if isinstance(x_or_x_list, list):
+            return self.call_nested(x_or_x_list, training=training)
+        elif hasattr(x_or_x_list, "shape") and hasattr(x_or_x_list, "dtype"):
+            return super().call(x_or_x_list, training=training)
+        else:
+            raise TypeError(f"Input must be a tensor or a list of tensors, but got {type(x_or_x_list)}")
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/block.py]: Let's double check that we really need this logic here.

File: paz/models/foundation/dinov2/layers/dino_head.py [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+        super().__init__(**kwargs)
+        self.units = units
+        self.use_bias = use_bias
+        self.kernel_initializer = initializers.get(kernel_initializer)
+        self.bias_initializer = initializers.get(bias_initializer)
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+
+        self.v = self.add_weight(
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/dino_head.py]: what is v?

File: paz/models/foundation/dinov2/layers/dino_head.py [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+            shape=(input_dim, self.units),
+            initializer=self.kernel_initializer,
+            trainable=True,
+            name="v",
+        )
+
+        self.g = self.add_weight(shape=(self.units,), initializer="ones", trainable=True, name="g")
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/dino_head.py]: what is g?

File: paz/models/foundation/dinov2/layers/dino_head.py [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+            kernel_initializer=initializers.TruncatedNormal(stddev=0.02),
+        )
+
+    def call(self, x):
+        x = self.mlp(x)
+        eps = 1e-6 if x.dtype == "float16" else 1e-12
+
+        x = ops.divide(x, ops.maximum(ops.sqrt(ops.sum(ops.square(x), axis=-1, keepdims=True)), eps))
+
+        x = self.last_layer(x)
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/dino_head.py]: Maybe this name can be different? It's clear that is the last layer. But maybe is not clear that is weight normalization dense?

File: paz/models/foundation/dinov2/layers/dino_head.py [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+
+        return x
+
+
+def _build_mlp(nlayers, in_dim, bottleneck_dim, hidden_dim=None, use_bn=False, bias=True):
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/dino_head.py]: complicated logic to read right away, let's try to make simpler to read by either making functions and full variable names.

File: paz/models/foundation/dinov2/layers/layer_scale.py [hunk 30]
```diff
@@ -0,0 +1,31 @@
+import keras
+import keras.ops as ops
+
+
+class LayerScale(keras.layers.Layer):
+    def __init__(self, dim, init_values=1e-5, inplace=False, dtype=None, **kwargs):
+        super().__init__(dtype=dtype, **kwargs)
+        self.dim = dim
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/layer_scale.py]: let's use full variable names dimension

File: paz/models/foundation/dinov2/layers/mlp.py [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+        )
+        self.drop_layer = keras.layers.Dropout(rate=drop, name="drop")
+
+    def call(self, x, training=None):
+        x = self.fc1(x)
+        x = self.act(x)
+        x = self.drop_layer(x, training=training)
+        x = self.fc2(x)
+        x = self.drop_layer(x, training=training)
+        return x
```
Comment (oarriaga) [inline paz/models/foundation/dinov2/layers/mlp.py]: I think we could replace this too be something like :

def block_MLP(x, some_params):
     x = Dense(params)(x)
     x = Activation(params)(x)
     ....

**Review by oarriaga — CHANGES_REQUESTED** [review]
